### ISSUE TITLE
Adjoint code in cylindrical coordinates

### ISSUE THREAD

**Issue by mochen4** [NanoComp/meep#1570]
It seems that adjoint code currently doesn't really support the cylindrical coordinates. For example, I get the error message ``Invalid fields::loop_in_chunks grid type ex for dimensions Cylindrical`` from https://github.com/NanoComp/meep/blob/8e2b0a050fd7e2d6d39019f76cdd3abcb73888c5/python/adjoint/optimization_problem.py#L194 in the forward run with the following setup. We probably need the different set of components for cylindrical coordinates.

```
...
resolution = 30

sr = 6
sz = 3
dimensions = mp.CYLINDRICAL
cell_size = mp.Vector3(sr, 0, sz)
pml_layers = [mp.PML(1.0)]
fcen = 1/1.55
width = 0.25
fwidth = width * fcen
source_center  = [3,0,-1]
source_size    = mp.Vector3(1,0,0)
src = mp.GaussianSource(frequency=fcen,fwidth=fwidth)
source = [mp.Source(src,component=mp.Ez,
                    center=source_center)]


design_region_resolution = 10
Nx = design_region_resolution
Ny = design_region_resolution
Si = mp.Medium(index=3.4)
SiO2 = mp.Medium(index=1.44)

design_variables = mp.MaterialGrid(mp.Vector3(Nx,Ny),SiO2,Si)
design_region = mpa.DesignRegion(design_variables,volume=mp.Volume(center=mp.Vector3(3,0,0), size=mp.Vector3(1, 0,1)))
geometry = [mp.Block(center=design_region.center, size=design_region.size, material=design_variables)]

sim = mp.Simulation(cell_size=cell_size, boundary_layers=pml_layers, geometry=geometry, sources=source,
                    eps_averaging=False, resolution=resolution, dimensions=dimensions,
                    m=2)


far_x = [mp.Vector3(0,0,30)]
NearRegions = [mp.Near2FarRegion(center=mp.Vector3(3,0 ,1),size=mp.Vector3(x=2), weight=+1)]
FarFields = mpa.Near2FarFields(sim, NearRegions ,far_x)
ob_list = [FarFields]

def J(alpha):
    return npa.abs(alpha[0,0,2])**2

opt = mpa.OptimizationProblem(
    simulation=sim,
    objective_functions=J,
    objective_arguments=ob_list,
    design_regions=[design_region],
    fcen=fcen,
    df = 0,
    nf = 1,
    decay_fields=[mp.Ez],
    decay_by=1e-4,
    maximum_run_time=1000
)

x0 = 11*np.random.rand(Nx*Ny) + 1
opt.update_design([x0])
```

**Comment by stevengj** [comment]
> We probably need the different set of components for cylindrical coordinates.

Presumably `[mp.Er,mp.Ep,mp.Ez]` should work?

**Comment by mochen4** [comment]
Using ``[mp.Er, mp.Ep, mp. Ez]``, the error now becomes ``meep: cannot require a er component in a 2D grid`` from https://github.com/NanoComp/meep/blob/8e2b0a050fd7e2d6d39019f76cdd3abcb73888c5/python/simulation.py#L2394 when setting up the adjoint run.

**Comment by stevengj** [comment]
That sounds like the adjoint run is being set up as Cartesian when it should be cylindrical?

**Comment by mochen4** [comment]
Thanks! I will find out where the setup is still Cartesian. As far as I can tell, the simulation itself is in cylindrical.

**Comment by mochen4** [comment]
I am still narrowing down the issue, but so far I find that after https://github.com/NanoComp/meep/blob/ef8f6fc950fdf7c5feb34705afe375e40d12c795/src/meepgeom.cpp#L2656 is called,  the ``get_material_gradient`` function is never called in ``material_grids_addgradient_point``.

**Comment by mochen4** [comment]
``min_max_loc`` and ``dims`` from https://github.com/NanoComp/meep/blob/ef8f6fc950fdf7c5feb34705afe375e40d12c795/src/meepgeom.cpp#L2630-L2631 appear to be inconsistent for cylindrical coordinates. ``min_max_loc`` views the cell as in x-z plane, but ``dims`` views the cell as in x-y plane. A superficial fix is to swap the values stored in ``dims`` for the corresponding axes after calling ``get_array_slice_dimensions`` if the simulation is cylindrical, but I don't know if we need to change anywhere else; and there probably are better ways to fix it. @smartalecH 

With the superficial fix, adjoint gradients are calculated. However, the fd_gradients are almost always zero right now. The adjoint gradients are also zero when the corresponding fd_gradients are zero. On the few instances where fd_gradients are nonzero, the adjoint is nonzero but the values don't match. The evaluation of the objective function itself is nonzero. I am not sure whether I have an incorrect setup that leads to gradients vanishing virtually everywhere.

**Comment by smartalecH** [comment]
Can you post a drawing of the simulation domain and optimization problem you are trying to solve (e.g. with `plot2D`)? Maybe some of the code too.

**Comment by mochen4** [comment]
The code is at the top of this page, just with ``m=0``. I also tried several variations but still couldn't locate the issue.

I don't think ``plot2D`` produces the right image; in particular, it has mysterious blocks and the cell goes from ``-sr/2`` to ``sr/2`` instead of ``0`` to ``sr``
![image](https://user-images.githubusercontent.com/25192039/119834916-5bf11a00-bece-11eb-8c37-29043bf8b6f4.png)


> Can you post a drawing of the simulation domain and optimization problem you are trying to solve (e.g. with `plot2D`)? Maybe some of the code too.

**Comment by smartalecH** [comment]
Oh ya... `plot2D` doesn't support cylindrical coordinates yet... although it still provides a good overview of the simulation problem.

Let's try to get a minimum working example (MWE) where the FD gradients are somewhat large (it will be much easier to debug).

I would span your design region from `r=0` to `r=sr`. Let's do the same for your source and far-field plane. For the objective, just maximize the power of a far-field point that's centered at `r=0` (the center of the cylindrical symmetry) but far in the z-direction.

i.e. something like [this](https://doi.org/10.1364/OE.403192):

![image](https://user-images.githubusercontent.com/24902086/119858012-0fafd500-bee2-11eb-9301-c7b7998db35a.png)

I'm not sure you want the `Ez` polarization though... maybe try `Er` (for the source, optimization stopping criteria, and far-field monitor). Perhaps one will give a larger gradient than the other...?

**Comment by mochen4** [comment]
I have been trying to figure out the issue, but the fd_gradients are still zero almost everywhere. I even tried a different setup based on the [zone plate](https://nbviewer.jupyter.org/github/NanoComp/meep/blob/master/python/examples/zone_plate.ipynb) example. However, a change in the structure doesn't change the far fields at all.

I also find that for different ``m`` values, the results of the forward simulation can be very different. For example, with ``Er`` source, at ``m=0``, only ``Ez`` component is appreciable at far-field; at ``m= -1``, ``Ez`` becomes negligible but ``Er, Ep, Hr, Hp`` become appreciable; at ``m= -2`` or larger, I get ``RuntimeError: meep: simulation fields are NaN or Inf
``. I don't think that is expected?

**Comment by stevengj** [comment]
If you are looking on the axis (at r=0), then by symmetry only Ez is nonzero for m=0 with an Er source, and for m=±1 only the (r,φ) components are nonzero at r=0.  For |m|>1 all of the field components must be zero at r=0.

**Comment by mochen4** [comment]
I switched to directional derivative for finite difference, and the results make more sense now. However, they still disagree with the adjoint gradient.

After carefully examining the code, I find a place in ``meepgeom.cpp`` that is potentially problematic. Just as ``dims`` and ``min_max_loc`` from https://github.com/NanoComp/meep/blob/fea1ce7127c51b454ddbb29bc5391c24f40cbc2e/src/meepgeom.cpp#L2666-L2667 used different conventions (i.e. one's coordinate is in x-y and the other in x-z), ``pb`` and ``sz`` in https://github.com/NanoComp/meep/blob/fea1ce7127c51b454ddbb29bc5391c24f40cbc2e/src/meepgeom.cpp#L2628-L2631 are also inconsistent. However, simply swapping  ``.y`` and ``.z`` for one of them still doesn't give the right result. It is actually worse becuase fd_gradient and adjoint gradient at least vanished at the same points previously, but it is no longer the case. On the other, ``sum(g_adjoint)`` is the same as before the swapping.

I am guessing there are other inconsistencies in ``meepgeom.cpp``, but I am a little confused about the code in general. Do you have any suggestions @smartalecH ? I also wonder what exactly is ``geometry_tree``? I tried to look it up but find no reference.

**Comment by smartalecH** [comment]
> However, they still disagree with the adjoint gradient.

How do they disagree? Is it just a scalar factor? Or something more complicated?

> However, simply swapping .y and .z for one of them still doesn't give the right result.

We might need a better way to calculate the dimensions in cylindrical coordinates. I admit it would be nice if the same routine worked for all coordinate systems, but let's first get the gradients to match up. Then we can refactor as needed.

Once we introduce the hybrid level-set functionality, this routine will have to change quite a bit anyway.

>  I also wonder what exactly is geometry_tree?

This is a data structure from [libctl](https://github.com/NanoComp/libctl) used to describe the geometry throughout the domain.

**Comment by stevengj** [comment]
Remember that `dV` is not a constant in cylindrical coordinates — it is scaled by 2πr.   So when you compute the inner product of the adjoint field with the forward field and dε/dp to find the `d/dp` derivative you might need to be careful.  (If this inner product were done with loop_in_chunks it would take care of this for you.)

**Comment by stevengj** [comment]
One easy check: make the design region very thin, all at a single radius, then if you are missing a 2πr factor it will just be a scale factor error.

**Comment by mochen4** [comment]
I noticed a couple more issues when using cylindrical coordinate:
1.  In my test case, points right on one of the edge got to https://github.com/NanoComp/meep/blob/fea1ce7127c51b454ddbb29bc5391c24f40cbc2e/src/meepgeom.cpp#L2589 so they aren't included. In comparison, points at all edges are included in 2D case. I am not sure how to fix this issue.
2.  ``dims`` in https://github.com/NanoComp/meep/blob/fea1ce7127c51b454ddbb29bc5391c24f40cbc2e/src/meepgeom.cpp#L2666-L2667 are ordered in ``z-x-y`` for some reason, i.e. ``dims[3*c]`` is the dimension in z for component c, etc. I simply swapped the dimensions for now. @smartalecH

**Comment by mochen4** [comment]
I noticed another place where ``get_array_slice_dimensions`` became problematic. After I fixed the order, the adjoint gradient only differs by a factor from the finite difference gradient.

Preliminary tests showed that the factor includes ``r`` for the ``DesginRegion`` and ``1/r`` for the ``Near2FarRegion``. I will do more tests, but so far, it appears that there is also an additional constant factor of 2 if and only if ``sim.m=0``.

**Comment by mochen4** [comment]
Here is the branch that almost works: https://github.com/mochen4/meep/tree/cyl_adj. In most cases, the finite difference and adjoint gradients match pretty well; in some cases, however, we need to run the simulations much longer, and although they still appear to be consistent, the adjoint and the finite difference gradients differ more.

**Comment by oskooi** [comment]
Closed by #1749.
